### DISCUSSION TITLE
n:n relationship with late bound classes

### DISCUSSION THREAD

**Discussion by sverheggen** [jordimontana82/fake-xrm-easy#568] [Q&A] [answered]
Hi there, 
I am trying to fake a N:N relationship. Not a manually created. I have an entity forexample student with the id studentid. The student can have multiple currencies. So I have a n:n from student to transactioncurrency. I am not using earlybound classes. 
So I created the following var student = new Entity("student", Guid.NewGuid); I created var currency = new Entity("transactioncurrency", Guid.NewGuid()); I create a new XrmFakeContext and addRelationship("student_currency_assocation",  IntersectEntity = "new_new_student_transactioncurrency",
                 Entity1LogicalName = transactioncurrency,
                 Entity1Attribute = "transactioncurrencyid",
                 Entity2LogicalName = "student",
                 Entity2Attribute = "new_student"
I initialize with these entities. And I have relationshipentity. 
The queryexpression starts at currency with an addLink to the new_new_student_transactioncurrency on studentid. 

When I run it I don't get any result. back. Hopefully somebody can help me with this struggle. Thanks in advanced for your help. Regards, Stefan

**Comment by jordimontana82** [comment, marked as answer]
Hi Stefan, without posting a full , executable, unit test would be near to impossible to troubleshoot, but in your example Entity2Attribute property seems to lack the "id" bit (i.e., new_studentid) would be the primary key attribute.  Hope this helps.

**Reply by sverheggen** [reply]
Hi Jordi, thanks for your quick response. I understand without the test it is difficult. I hoped there was an example how to set it up that works. But I created one: 
 [Fact]
        public void TestManyToMany()
        {
            // arrange
            var context = new XrmFakedContext();
            var currencyEntity = new Entity("transactioncurrency", Guid.NewGuid());
            var studentEntity = new Entity("new_student", Guid.NewGuid());

            var relationEntity = new Entity("rav_rav_ole_transactioncurrency", Guid.NewGuid());
            relationEntity.Attributes.Add(new KeyValuePair<string, object>("transactioncurrency", currencyEntity.Id));
            relationEntity.Attributes.Add(new KeyValuePair<string, object>("new_studentid", studentEntity.Id));

            context.AddRelationship("new_new_student_transactioncurrency", new XrmFakedRelationship
            {
                IntersectEntity = "new_new_student_transactioncurrency",
                Entity1LogicalName = currencyEntity.LogicalName,
                Entity1Attribute = "transactioncurrencyid",
                Entity2LogicalName = studentEntity.LogicalName,
                Entity2Attribute = "new_studentid"
            });

            context.Initialize(new List<Entity> { currencyEntity, studentEntity, relationEntity });

            // act
            var query = new QueryExpression("transactioncurrency");
            query.TopCount = 1;
            query.Criteria.AddCondition("transactioncurrencyid", ConditionOperator.Equal, currencyEntity.Id);
            query.ColumnSet = new ColumnSet("transactioncurrencyid");

            var studentRelation = query.AddLink("new_new_student_transactioncurrency", "transactioncurrencyid", "transactioncurrencyid");

            studentRelation.LinkCriteria.AddCondition("new_studentid", ConditionOperator.Equal, studentEntity.Id);

            var result = context.GetOrganizationService().RetrieveMultiple(query);

            // assert

        }

If this test is run the result with contain 0 entities and I believe it should contain 1. I don't now if the relationentity part is necessary. Running with or without doesn't make a difference. Hopefully you can help me out with this example. For the rest it is a really great piece of software able to mock the CRM System. Thanks for that. Also thanks for helping me out. 
Regards, Stefan

**Reply by BetimBeja** [reply]
as with the `new_studentid` this line
```csharp
  relationEntity.Attributes.Add(new KeyValuePair<string, object>("transactioncurrency", currencyEntity.Id));
```
should be like this
```csharp
  relationEntity.Attributes.Add(new KeyValuePair<string, object>("transactioncurrencyid", currencyEntity.Id));
```

**Reply by sverheggen** [reply]
Thanks for your help looking with me. It was the two letters (id).

**Reply by jordimontana82** [reply]
Thanks Stefan
